features/net/network-socket/nsapi_dns.h

```diff
@@ -0,0 +1,103 @@
+/* nsapi_dns.h
+ * Original work Copyright (c) 2013 Henry Leinen (henry[dot]leinen [at] online [dot] de)
+ * Modified work Copyright (c) 2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef NSAPI_DNS_H
+#define NSAPI_DNS_H
+
+#include "nsapi_types.h"
+#include "network-socket/NetworkStack.h"
+
+
+/** Query a domain name server for an IP address of a given hostname
+ *
+ *  @param stack    Network stack as target for DNS query
+ *  @param addr     Destination for the host address
+ *  @param host     Hostname to resolve
+ *  @param version  IP version to resolve (defaults to NSAPI_IPv4)
+ *  @return         0 on success, negative error code on failure
+ *                  NSAPI_ERROR_DNS_FAILURE indicates the host could not be found
+ */
+NSAPI_C_LINKAGE
+int nsapi_dns_query(nsapi_stack_t *stack, nsapi_addr_t *addr,
+        const char *host, nsapi_version_t version);
+
+#ifdef __cplusplus
```

Comment:
either provide that abstraction (is it worth doing in this case?) or split this header file or logically split this to C only and then C++ (or other way around) in this file, might improve readability? I dont have strong preference, can be as it is, just not  that clear when I saw it the first time. my 5 cents.

does this file compile if included from C? Because it includes NetworkStack that is pure C++?

> The original goal was to avoid duplicated function definitions.
> If we don't care about duplicated code

I don't think there would be anything duplicated, or? C would declare C API, C++ would do for C++ and include C header file (if we want to expose C functions). The implementation can stay in the cpp file, but providing two headers. Anyway, this is a case only if we go that road to split, we could another, so lets decide which one is user friendly.